### PR TITLE
Integrate cache daemon into the CLI.

### PR DESCRIPTION
**Pull request by mefyl**

### CODE REVIEW

File: bin/cache_daemon.ml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  [ `S "DESCRIPTION"
+  ; `P
+      {|Dune allows to share build artifacts between workspaces.
+        $(b,dune cache-daemon) is a daemon that runs in the background
+        and manages this shared cache. For instance, it makes sure that it
+        does not grow too big and try to maximise sharing between the various
+        workspace that are using the shared cache.|}
+  ; `P
+      {|The daemon is automatically started by Dune when the shared cache is
+                                                             enabled. You do not need to run this command manually.|}
```
Comment (diml) [inline bin/cache_daemon.ml]: We can delete a few spaces here

File: bin/cache_daemon.ml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+         value
+         & opt path_conv (Dune_memory.default_root ())
+         & info ~docv:"PATH" [ "root" ] ~doc:"Root of the dune cache")
+     in
+     try
+       match mode with
+       | Some Start -> `Ok (start ~exit_no_client ~foreground ~port_path ~root)
+       | Some Stop -> `Ok (stop ~port_path)
+       | None -> `Help (`Pager, Some name)
+     with Failure msg -> `Error (false, msg)
```
Comment (diml) [inline bin/cache_daemon.ml]: We shouldn't catch this error. There is already a catch-all error handler in `main.ml`

Comment (mefyl) [reply bin/cache_daemon.ml]: Reading the code I do agree, but computer says no: ```Error: exception Failure("too many retries (100): waiting for daemon to stop (PID 10860)")
Backtrace:
Raised at file "src/stdune/result.ml", line 17, characters 15-22
Called from file "bin/cache_daemon.ml", line 124, characters 24-41
Called from file "vendor/cmdliner/src/cmdliner_term.ml", line 25, characters 19-24
Called from file "vendor/cmdliner/src/cmdliner.ml", line 30, characters 14-21
Called from file "vendor/cmdliner/src/cmdliner.ml", line 146, characters 9-16
Called from file "vendor/cmdliner/src/cmdliner.ml", line 176, characters 18-36
Called from file "vendor/cmdliner/src/cmdliner.ml", line 312, characters 20-46
Called from file "bin/main.ml", line 198, characters 10-51

I must not segfault.  Uncertainty is the mind-killer.  [...]```

Comment (mefyl) [reply bin/cache_daemon.ml]: Ah wait, that's the catchall. But I don't want a frightening assertion message, I just want to print out "couldn't stop the daemon after X tries".

Comment (diml) [reply bin/cache_daemon.ml]: What about using `User_error.raise` in `retry` rather than returning `Error (Failure ...)`?

Comment (mefyl) [reply bin/cache_daemon.ml]: Absolutely ; done.

**Pull request merged**